src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+#include "set_union.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "storage/chunk.hpp"
+#include "storage/reference_column.hpp"
+#include "storage/table.hpp"
+#include "types.hpp"
+
+/**
+ * ### SetUnion implementation
+ * The SetUnion Operator turns each input table into a ReferenceMatrix.
+ * The rows in the ReferenceMatrices need to be sorted in order for them to be merged, that is,
+ * performing the SetUnion  operation.
+ * Since sorting a multi-column matrix by rows would require a lot of value-copying, for each ReferenceMatrix, a
+ * VirtualPosList is created.
+ * Each element of this VirtualPosList references a row in a ReferenceMatrix by index. This way, if two values need to
+ * be swapped while sorting, only two indices need to swapped instead of a RowIDs for each column in the
+ * ReferenceMatrices.
+ * Using a implementation derived from std::set_union, the two virtual pos lists are merged into the result table.
+ *
+ *
+ * ### About ReferenceMatrices
+ * The ReferenceMatrix consists of N rows and C columns of RowIDs.
+ * N is the same number as the number of rows in the input table.
+ * Each of the C column can represent 1..S columns in the input table. All rows represented by a ReferenceMatrix-Column
+ * contain the same PosList in each Chunk.
+ *
+ * The ReferenceMatrix of a StoredTable will only contain one column, the ReferenceMatrix of the result of a 3 way Join
+ * will contain 3 columns
+ *
+ *
+ * ### TODO(anybody) for performance improvements
+ * Instead of using a ReferenceMatrix, consider using a linked list of RowIDs for each row. Since most of the sorting
+ *      will depend on the leftmost column, this way most of the time no remote memory would need to be accessed
+ *
+ * The sorting, which is the most expensive part of this operator, could probably be parallelized.
+ */
+
+namespace {
+
+// See doc above
+using ReferenceMatrix = std::vector<opossum::PosList>;
+
+using VirtualPosList = std::vector<size_t>;
+
+/**
+ * Comparator for performing the std::sort() of a virtual pos list
+ */
+struct VirtualPosListCmpContext {
```

Comment:
Is this a "context" or a "comparator"? Also, it should be clear in the name that it is actually a less-than comparator

Reply:
All Comparators in C++ are less than, so this would feel out of place. 
It's both a Context (since it holds a ref to the ReferenceMatrix) and a Comparator. Is this a problem?

Reply:
Ok, didn't know about "compare" in c++.
When I see `Context` in our code, I usually think about the visitors. For me, a context is something that holds information that is meant to be passed around. `reference_matrix_column` here never leaves the comparator.